### PR TITLE
build wheel in github actions

### LINKED ISSUE #154: Distribute some wheels on pypi
**[Original report](https://bitbucket.org/blais/beancount/issue/154) by rkhwaja (Bitbucket: [rkhwaja](https://bitbucket.org/rkhwaja), GitHub: [rkhwaja](https://github.com/rkhwaja)).**

----------------------------------------

I'm particuarly interested in python 3.6 on windows

### PR DESCRIPTION
**Pull request by trim21**
looks like it's blocked by #536, workflow shoule work fine if #536 is fixed, see https://test.pypi.org/project/beancount/#files

You need to add a github actions secrets named `PYPI_TOKEN`

This PR also add a workflow to test if it compile fine on windows https://github.com/beancount/beancount/pull/537/files#diff-6d1ed6f2f15040f1022bb9655a581b0d

close #154

### CODE REVIEW

File: .github/workflows/pypi-wheel.yaml
Comment (dnicolodi) [inline .github/workflows/pypi-wheel.yaml]: I don't think this is required, is it? `setup.py` already adds the `YY_NO_UNISTD_H` define on windows. Is the `--compiler=msvc` option required?

File: .github/workflows/pypi-wheel.yaml
Comment (dnicolodi) [inline .github/workflows/pypi-wheel.yaml]: Missing return at the end of the file. Only a cosmetic issue, but while we are here...

File: .github/workflows/pypi-wheel.yaml
Comment (dnicolodi) [inline .github/workflows/pypi-wheel.yaml]: What is this for?

File: .github/workflows/pypi-wheel.yaml
Comment (dnicolodi) [inline .github/workflows/pypi-wheel.yaml]: Can you please drop the empty lines? At least for consistency with the Linux recipe below

File: .github/workflows/windows-ci.yaml
Comment (dnicolodi) [inline .github/workflows/windows-ci.yaml:11]: I don't think there is any need compile-check on all supported Python version, our use of the Python CAPI is very basic and I don't expect any issue. Although, if it is fast I don't see a reason why to drop it...

File: .github/workflows/windows-ci.yaml
Comment (dnicolodi) [inline .github/workflows/windows-ci.yaml]: Can you please drop the empty lines in this recipe too?

File: .github/workflows/windows-ci.yaml
Comment (dnicolodi) [inline .github/workflows/windows-ci.yaml]: I don't see a good reason why we need to save the wheel either. Is not like they may be useful to reproduce a test error as no tests are run on the CI anyway.

File: .github/workflows/windows-ci.yaml
Comment (dnicolodi) [inline .github/workflows/windows-ci.yaml]: I don't see the point of creating a wheel if the only thing we are checking is that the code compiles. I would drop this and the `pip install wheel` above.

**Review by dnicolodi — CHANGES_REQUESTED** [review]
I am not sure about publishing wheels for macOS and Linux where compiling beancount is trivial, however I think it would be nice to have them for Windows. Compiling the code for PRs on Windows is definitely a nice addition. Can you please comment or fix my remarks and squash the commits in two patches: one removing the AppVeyor rule files and one adding the Github Actions rule files? Also, the PR should be for the v2 branch and not the master branch. Thank you!

File: .github/workflows/pypi-wheel.yaml
Comment (trim21) [reply .github/workflows/pypi-wheel.yaml]: Most part of steps on windows are migrated from orginal `appveyor.yml` file, and it's deleted in this pr. I'm not sure if it's required.

I'll test it later on #543 or #538 to see if it's necessary.

File: .github/workflows/pypi-wheel.yaml
Comment (trim21) [reply .github/workflows/pypi-wheel.yaml]: Sorry, debug. I'll remove it.

File: .github/workflows/windows-ci.yaml
Comment (trim21) [reply .github/workflows/windows-ci.yaml]: What orginal `appveyor.yml` did, too.

File: .github/workflows/windows-ci.yaml
Comment (trim21) [reply .github/workflows/windows-ci.yaml]: add a step `twine check dist/*` to check wheel file.

**Comment by blais** [comment]
Let me know when you guys are both satisfied with this new config, happy to merge it anytime.

File: .github/workflows/pypi-wheel.yaml
Comment (dnicolodi) [reply .github/workflows/pypi-wheel.yaml]: I'm sure it is not necessary

**Review by dnicolodi — CHANGES_REQUESTED** [review]
Something went wrong with your last push. Can you please squash the commits and rebase on the v2 branch?

**Comment by trim21** [comment]
@dnicolodi I'm rebasing it, I just changed the pr base branch.

**Comment by dnicolodi** [comment]
> Let me know when you guys are both satisfied with this new config, happy to merge it anytime.

@blais, can you please merge #543 in the meanwhile? I setup AppVeyor CI on my beancount fork. That PR fixes the compilation on Windows and I like that solution more than the one proposed by @skobes. Once you do that, you can also close the #538 PR.

File: .github/workflows/pypi-wheel.yaml
Comment (dnicolodi) [inline .github/workflows/pypi-wheel.yaml]: Now this step can be removed, or at least the conditional.

File: .github/workflows/pypi-wheel.yaml
Comment (dnicolodi) [inline .github/workflows/pypi-wheel.yaml]: Don't we need to build the wheels before uploading them with twine? I think a
```
      - run: python setup.py bdist_wheel
```
line got lost somewhere (or I don't understand how this is supposed to work).

**Review by dnicolodi — CHANGES_REQUESTED** [review]
I think we are close.

File: .github/workflows/pypi-wheel.yaml
Comment (trim21) [reply .github/workflows/pypi-wheel.yaml]: They are compiled by `uses: RalfG/python-wheels-manylinux-build@v0.3.1`

If we just compile it by `bdist_wheel` on linux we will get wheels with`linux` tag and it's not supported by pypi(why we upload `dist/*manylinux*`). We compile it with manylinux then we get wheels with `manylinux` tag.

Or maybe we just remove this job considering compiling C extension is very easy on linux.

**Comment by blais** [comment]
@dnicolodi #543 merged
Let me know 
(I'm working on returns computation, I'm around today)

**Comment by dnicolodi** [comment]
@blais I think the CI job compiling the code on Windows is a nice thing to have, waiting for having also the tests run on Windows. The wheels for Windows are also a good thing to have, as it is a bit harder to compile the code there. I don't know about the macOS ones. We can probably do without the Linux ones. I'll let you decide on that.

**Comment by trim21** [comment]
I tried add testing on Windows, but it seems that not all tests is passed…

Simply `pytest beancount`, not sure I did it right.

**Comment by dnicolodi** [comment]
> I tried add testing on Windows, but it seems that not all tests is passed…

A lot of the test failures are due to the tests assuming that `/` is the path separator, however that is not true on Windows. Other failures are due to the handling of temporary files on Windows. I started the work to fix the tests, but fixing the first issue is tedious and I have very little time and even less motivation.

**Comment by trim21** [comment]
maybe enable them part by part. I can help with it (only python, I know nothing about c or cpp)🤔

File: .github/workflows/pypi-wheel.yaml
Comment (dnicolodi) [reply .github/workflows/pypi-wheel.yaml]: I think having Linux wheels is not worth the trouble of maintaining the infrastructure running. By the fact that building them requires some non trivial steps I conclude that building Linux wheels is not as straightforward as on other platforms. Also, we have done without wheels on Linux for as long as beancount existed and I haven't heard too many complaints. Can you please remove this job?

**Comment by dnicolodi** [comment]
Can you please squash the two commits into one? I think after that we can merge this. Thank you

**Comment by trim21** [comment]
why not choose “squash and commit” when merging pr......





------------------ Original ------------------

**Comment by dnicolodi** [comment]
> why not choose “squash and commit” when merging pr

Because I am not the one merging the PR and I want to make things as straightforward as possible for Martin. Thank you for your cooperation

**Pull request merged**
